ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/events/filesystem/ConstraintEventsIterator.java

```diff
@@ -0,0 +1,94 @@
+/*
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.exec.repl.bootstrap.events.filesystem;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Iterator;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.parse.EximUtil;
+import org.apache.hadoop.hive.ql.parse.ReplicationSemanticAnalyzer;
+
+public class ConstraintEventsIterator implements Iterator<FSConstraintEvent> {
+  private FileStatus[] dbDirs;
+  private int currentDbIndex;
+  private FileStatus[] constraintFiles = null;
+  private int currentConstraintIndex;
+  private FileSystem fs;
+  private Path path;
+
+  public ConstraintEventsIterator(String dumpDirectory, HiveConf hiveConf) throws IOException {
+    path = new Path(dumpDirectory);
+    fs = path.getFileSystem(hiveConf);
+  }
+
+  private FileStatus[] listConstraintFilesInDBDir(FileSystem fs, Path dbDir) {
+    try {
+      return fs.listStatus(new Path(dbDir, ReplicationSemanticAnalyzer.CONSTRAINTS_ROOT_DIR_NAME));
+    } catch (FileNotFoundException e) {
+      return new FileStatus[]{};
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (dbDirs == null) {
+      try {
+        dbDirs = fs.listStatus(path, EximUtil.getDirectoryFilter(fs));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      currentDbIndex = 0;
+      if (dbDirs.length != 0) {
+        currentConstraintIndex = 0;
+        constraintFiles = listConstraintFilesInDBDir(fs, dbDirs[0].getPath());
+      }
+    }
+    if (currentDbIndex < dbDirs.length && currentConstraintIndex < constraintFiles.length) {
+      return true;
+    }
+    while (currentDbIndex < dbDirs.length && currentConstraintIndex == constraintFiles.length) {
```

Comment:
Shall explicitly mark the conditions boundary with parenthesis.

Reply:
Those are single condition, right?

Reply:
Yup, but just suggesting to put conditions inside parenthesis for better readability.
((currentDbIndex < dbDirs.length) && (currentConstraintIndex == constraintFiles.length))